libs/@local/graph/store/src/filter/mod.rs

```diff
@@ -423,122 +423,120 @@ impl<'p> Filter<'p, Entity> {
+                (None, Effect::Permit) => blank_permit = true,
+                (None, Effect::Forbid) => return Self::Any(Vec::new()), // Blank forbid = deny all
                 (Some(resource), Effect::Permit) => {
+                    // Non-optimizable permits
                     permits.push(Self::for_resource_constraint(resource, actor_id));
                 }
                 (Some(resource), Effect::Forbid) => {
+                    // All forbids
                     forbids.push(Self::for_resource_constraint(resource, actor_id));
                 }
-                (None, Effect::Permit) => {
-                    blank_permit = true;
-                }
-                (None, Effect::Forbid) => {
-                    // A forbid without a filter means we never allow access.
-                    return Self::Any(Vec::new());
-                }
             }
         }
+
+        // Add optimized entity permits if any
+        match optimization_data.permitted_entity_uuids.len() {
+            0 => {}
+            1 => {
+                let id = optimization_data.permitted_entity_uuids[0];
+                permits.push(Self::Equal(
+                    Some(FilterExpression::Path {
+                        path: EntityQueryPath::Uuid,
+                    }),
+                    Some(FilterExpression::Parameter {
+                        parameter: Parameter::Uuid(id.into()),
+                        convert: None,
+                    }),
+                ));
+            }
+            _ => {
+                // Use the Vec directly for the IN clause
+                permits.push(Self::In(
+                    FilterExpression::Path {
+                        path: EntityQueryPath::Uuid,
+                    },
+                    ParameterList::EntityUuids(&optimization_data.permitted_entity_uuids),
+                ));
+            }
+        }
+
+        // Add optimized web ID permits if any
+        match optimization_data.permitted_web_ids.len() {
+            0 => {}
+            1 => {
+                let web_id = optimization_data.permitted_web_ids[0];
+                permits.push(Self::Equal(
+                    Some(FilterExpression::Path {
+                        path: EntityQueryPath::WebId,
+                    }),
+                    Some(FilterExpression::Parameter {
+                        parameter: Parameter::Uuid(web_id.into()),
+                        convert: None,
+                    }),
+                ));
+            }
+            _ => {
+                // Use the Vec directly for the IN clause
+                permits.push(Self::In(
+                    FilterExpression::Path {
+                        path: EntityQueryPath::WebId,
+                    },
+                    ParameterList::WebIds(&optimization_data.permitted_web_ids),
+                ));
+            }
+        }
+
+        // Apply the same combination logic as for_policies
         if blank_permit {
```

Comment:
I feel like we're doing a lot of additional work even if we just end up with a blank permit. (not that it is blocking, just what I observed)